File: samples/client/petstore/android-java/src/main/java/com/wordnik/client/ApiInvoker.java
package com.wordnik.client;


import com.fasterxml.jackson.core.JsonGenerator.Feature;
import com.fasterxml.jackson.databind.*;
import com.fasterxml.jackson.annotation.*;
import com.fasterxml.jackson.databind.annotation.JsonSerialize;

import org.apache.http.*;
import org.apache.http.client.*;
import org.apache.http.client.methods.*;
import org.apache.http.entity.StringEntity;
import org.apache.http.impl.client.*;
import org.apache.http.util.EntityUtils;

import java.io.File;
import java.net.URLEncoder;

import java.util.Map;
import java.util.HashMap;
import java.util.List;
import java.io.IOException;
import java.io.UnsupportedEncodingException;

public class ApiInvoker {
  private static ApiInvoker INSTANCE = new ApiInvoker();
  private Map<String, String> defaultHeaderMap = new HashMap<String, String>();

  private HttpClient client = null;
  public static ApiInvoker getInstance() {
    return INSTANCE;
  }

  public void addDefaultHeader(String key, String value) {
    defaultHeaderMap.put(key, value);
  }

  public String escapeString(String str) {
    return str;
  }

  public static Object deserialize(String json, String containerType, Class cls) throws ApiException {
    try{
      if("List".equals(containerType)) {
        JavaType typeInfo = JsonUtil.getJsonMapper().getTypeFactory().constructCollectionType(List.class, cls);
        List response = (List<?>) JsonUtil.getJsonMapper().readValue(json, typeInfo);
        return response;
      }
      else if(String.class.equals(cls)) {
        if(json != null && json.startsWith("\"") && json.endsWith("\"") && json.length() > 1)
          return json.substring(1, json.length() - 2);
        else 
          return json;
      }
      else {
        return JsonUtil.getJsonMapper().readValue(json, cls);
      }
    }
    catch (IOException e) {
      throw new ApiException(500, e.getMessage());
    }
  }

  public static String serialize(Object obj) throws ApiException {
    try {
      if (obj != null) 
        return JsonUtil.getJsonMapper().writeValueAsString(obj);
      else 
        return null;
    }
    catch (Exception e) {
      throw new ApiException(500, e.getMessage());
    }
  }

  public String invokeAPI(String host, String path, String method, Map<String, String> queryParams, Object body, Map<String, String> headerParams, String contentType) throws ApiException {
    HttpClient client = getClient(host);

    StringBuilder b = new StringBuilder();
    for(String key : queryParams.keySet()) {
      String value = queryParams.get(key);
      if (value != null){
        if(b.toString().length() == 0)
          b.append("?");
        else
          b.append("&");
        b.append(escapeString(key)).append("=").append(escapeString(value));
      }
    }
    String url = host + path + b.toString();

    HashMap<String, String> headers = new HashMap<String, String>();

    for(String key : headerParams.keySet()) {
      headers.put(key, headerParams.get(key));
    }
    
    for(String key : defaultHeaderMap.keySet()) {
      if(!headerParams.containsKey(key)) {
        headers.put(key, defaultHeaderMap.get(key));
      }
    }
    headers.put("Accept", "application/json");

    HttpResponse response = null;
    try{
      if("GET".equals(method)) {
        HttpGet get = new HttpGet(url);
        get.addHeader("Accept", "application/json");
        for(String key : headers.keySet()) {
          get.setHeader(key, headers.get(key));
        }
        response = client.execute(get);
      }
      else if ("POST".equals(method)) {
        HttpPost post = new HttpPost(url);
        post.setHeader("Content-Type", contentType);
        post.setEntity(new StringEntity(serialize(body), "UTF-8"));
        for(String key : headers.keySet()) {
          post.setHeader(key, headers.get(key));
        }
        response = client.execute(post);
      }
      else if ("PUT".equals(method)) {
        HttpPut put = new HttpPut(url);
        if(body != null) {
          put.setHeader("Content-Type", contentType);
          put.setEntity(new StringEntity(serialize(body), "UTF-8"));
        }
        for(String key : headers.keySet()) {
          put.setHeader(key, headers.get(key));
        }
        response = client.execute(put);
      }
      else if ("DELETE".equals(method)) {
        HttpDelete delete = new HttpDelete(url);
        for(String key : headers.keySet()) {
          delete.setHeader(key, headers.get(key));
        }
        response = client.execute(delete);
      }

      int code = response.getStatusLine().getStatusCode();
      String responseString = null;
      if(code == 204) 
        responseString = "";
      else if(code >= 200 && code < 300) {
        if(response.getEntity() != null) {
          HttpEntity resEntity = response.getEntity();
          responseString = EntityUtils.toString(resEntity);
        }
      }
      else {
        if(response.getEntity() != null) {
          HttpEntity resEntity = response.getEntity();
          responseString = EntityUtils.toString(resEntity);
        }
        else
          responseString = "no data";
        throw new ApiException(code, responseString);
      }
      return responseString;
    }
    catch(IOException e) {
      throw new ApiException(500, e.getMessage());
    }
  }

  private HttpClient getClient(String host) {
    if(client == null)
      client = new DefaultHttpClient();
    return client;
  }
}

File: samples/client/petstore/csharp/src/Com/Wordnik/Petstore/ApiInvoker.cs
  using System;
  using System.Collections.Generic;
  using System.IO;
  using System.Linq;
  using System.Net;
  using System.Text;
  using Newtonsoft.Json;

  namespace Com.Wordnik.Petstore {
    public class ApiInvoker {
      private static readonly ApiInvoker _instance = new ApiInvoker();
      private Dictionary<String, String> defaultHeaderMap = new Dictionary<String, String>();

      public static ApiInvoker GetInstance() {
        return _instance;
      }
      
      public void addDefaultHeader(string key, string value) {
         defaultHeaderMap.Add(key, value);
      }

      public string escapeString(string str) {
        return str;
      }

      public static object deserialize(string json, Type type) {
        try
        {
            return JsonConvert.DeserializeObject(json, type);
        }
        catch (IOException e) {
          throw new ApiException(500, e.Message);
        }

      }

      public static string serialize(object obj) {
        try
        {
            return obj != null ? JsonConvert.SerializeObject(obj) : null;
        }
        catch (Exception e) {
          throw new ApiException(500, e.Message);
        }
      }

      public string invokeAPI(string host, string path, string method, Dictionary<String, String> queryParams, object body, Dictionary<String, String> headerParams) {
        var b = new StringBuilder();
        
        foreach (var queryParamItem in queryParams)
        {
            var value = queryParamItem.Value;
            if (value == null) continue;
            b.Append(b.ToString().Length == 0 ? "?" : "&");
            b.Append(escapeString(queryParamItem.Key)).Append("=").Append(escapeString(value));
        }

        var querystring = b.ToString();

          host = host.EndsWith("/") ? host.Substring(0, host.Length - 1) : host;

          var client = WebRequest.Create(host + path + querystring);
          client.ContentType = "application/json";
          client.Method = method;

          foreach (var headerParamsItem in headerParams)
          {
              client.Headers.Add(headerParamsItem.Key, headerParamsItem.Value);
          }
          foreach (var defaultHeaderMapItem in defaultHeaderMap.Where(defaultHeaderMapItem => !headerParams.ContainsKey(defaultHeaderMapItem.Key)))
          {
              client.Headers.Add(defaultHeaderMapItem.Key, defaultHeaderMapItem.Value);
          }
    
          switch (method)
          {
              case "GET":
                  break;
              case "POST":
              case "PUT":
              case "DELETE":
                  var swRequestWriter = new StreamWriter(client.GetRequestStream());
                  swRequestWriter.Write(serialize(body));
                  swRequestWriter.Close();
                  break;
              default:
                  throw new ApiException(500, "unknown method type " + method);         
          }
              var webResponse = (HttpWebResponse) client.GetResponse();
              if (webResponse.StatusCode != HttpStatusCode.OK) throw new ApiException((int) webResponse.StatusCode, webResponse.StatusDescription);

              var responseReader = new StreamReader(webResponse.GetResponseStream());
              var responseData = responseReader.ReadToEnd();
              responseReader.Close();
              return responseData;
      }

    }
  }

